pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+        full_stack = stack()
+
+        # 2. discard frames whose filename begins with "<"
+        frame_infos = list(filter(StackFrame._frame_info_not_py_crud, full_stack))
+
+        def _make_stack_frames(fis):
+            return [
+                frame
+                for f in fis
+                if (frame := StackFrame._init_or_drop(self, f, full_stack))
+            ]
+
+        if self._keep_all_debugging or len(frame_infos) <= 1:
+            self._frames = _make_stack_frames(frame_infos)
+            return
+
+        # 3. start the best frame search right after where NatalStackFrame() was constructed
+        # For more details see the unit test:
+        # tests/unit/sourcemap_monkey_unit_test.py::test_frame_info_is_right_before_core_last_drop_idx
+        i = 2  # formerly this was `last_drop_idx = 1; i = last_drop_idx + 1`
+
+        # 4. fast forward the right bound until we're out of pyteal-library code
+        # This sets last_keep_idx to the first frame index which isn't pyteal
+        while i < len(frame_infos) and StackFrame._frame_info_is_pyteal(frame_infos[i]):
+            i += 1
+        last_keep_idx = i
+
+        # 5. if the pyteal-library exit point was an import, the expression was
+        # generated by pyteal itself. So let's back up and look for a "# T2PT*" comment
+        # which will give us a clue for what to do with this expression
+        if StackFrame._frame_info_is_pyteal_import(frame_infos[last_keep_idx]):
+            found = False
+            i = -1
+            for i in range(last_keep_idx - 1, -1, -1):
+                if StackFrame._frame_info_compiler_generated(frame_infos[i]):
+                    found = True
+                    break
+
+            if found and i >= 0:
+                last_keep_idx = i
+
+        # 6. Keep only the last frame in the list. We maintain _as_ a list
+        # since in the case of `self._debug == True`, we'd like access to the full list.
+        # TODO: this is likely obsolete since full_stack is available on the PyTealFrame object when debugging
+        frame_infos = frame_infos[last_keep_idx : last_keep_idx + 1]
+
+        # 7. we finish by constructing a list[StackFrame] from our one remaining frame_info
+        self._frames = _make_stack_frames(frame_infos)
+
+    def __len__(self) -> int:
+        return len(self._frames)
+
+    def best(self) -> StackFrame:
+        """
+        Return the best guess as to the user-authored birthplace of the
+        associated StackFrame's
+        """
+        assert (
+            self._frames
+        ), f"expected to have some frames but currently {self._frames=}"
+        return self._frames[-1]
+
+    def __repr__(self) -> str:
+        return f"{'C' if self._compiler_gen else 'U'}{self._frames}"
+
+    def nodes(self) -> list[AST | None]:
+        return [f.node for f in self._frames]
+
+    @classmethod
+    def _walk_asts(cls, func: Callable[["Expr"], None], *exprs: "Expr") -> None:  # type: ignore
```

Comment:
Right. During our yesterday's meeting @jasonpaulos brought up some possible issues with this approach. In particular, it can only visit the types of `Expr`'s that it knows about and we need to avoid going into infinite loops.

Reply:
In the router improvements PR #676 , the walk is exited when the _same_ `stack_frames` object that is being set is encountered. That should avoid infinite loops. I can bring that approach in, along with a unit test that has an infinite loop when the criteria isn't followed.